pkg/server/structlogging/hot_ranges_log.go

```diff
@@ -103,57 +130,122 @@ func (s *hotRangesLoggingScheduler) startJob(ctx context.Context, stopper *stop.
-	TelemetryHotRangesStatsInterval.SetOnChange(&s.st.SV, func(ctx context.Context) {
-		intervalChangedChan <- struct{}{}
-	})
-
-	ticker := time.NewTicker(TelemetryHotRangesStatsInterval.Get(&s.st.SV))
-
 	for {
+		checkInterval := CheckInterval
+		if s.multiTenant {
+			checkInterval *= 5
+		}
 		select {
 		case <-stopper.ShouldQuiesce():
 			return nil
 		case <-ctx.Done():
 			return nil
-		case <-ticker.C:
+		case <-time.After(checkInterval):
 			s.maybeLogHotRanges(ctx, stopper)
-		case <-intervalChangedChan:
-			ticker.Reset(TelemetryHotRangesStatsInterval.Get(&s.st.SV))
+		case <-TestLoopChannel:
+			continue
 		}
 	}
 }
 
 // maybeLogHotRanges is a small helper function which couples the
 // functionality of checking whether to log and logging.
 func (s *hotRangesLoggingScheduler) maybeLogHotRanges(ctx context.Context, stopper *stop.Stopper) {
-	if s.shouldLog() {
+	if s.shouldLog(ctx) {
 		s.logHotRanges(ctx, stopper)
+		s.lastLogged = timeutil.Now()
 	}
 }
 
-// shouldLog checks the below conditions to see whether it should emit logs.
-//   - Is the cluster setting server.telemetry.hot_ranges_stats.enabled true?
-func (s *hotRangesLoggingScheduler) shouldLog() bool {
-	return TelemetryHotRangesStatsEnabled.Get(&s.st.SV)
+// shouldLog is a passthrough to the below should log implementation,
+// where the requisite information is pulled out so that the
+// rules can be applied.
+//
+// The enabled check is pulled upwards so that we don't have to
+// make the rpc call if it's disabled.
+func (s *hotRangesLoggingScheduler) shouldLog(ctx context.Context) bool {
+	enabled := TelemetryHotRangesStatsEnabled.Get(&s.st.SV)
+	if !enabled {
+		return false
+	}
+
+	logInterval := TelemetryHotRangesStatsInterval.Get(&s.st.SV)
+	resp, err := s.getHotRanges(context.Background(), true)
+	if err != nil {
+		log.Warningf(ctx, "failed to get hot ranges: %s", err)
+		return false
+	}
+
+	return shouldLog(
+		enabled,
+		resp.Ranges,
+		s.lastLogged,
+		logInterval,
+		&timeutil.DefaultTimeSource{},
+	)
 }
 
-// logHotRanges collects the hot ranges from this node's status server and
-// sends them to the TELEMETRY log channel.
-func (s *hotRangesLoggingScheduler) logHotRanges(ctx context.Context, stopper *stop.Stopper) {
-	req := &serverpb.HotRangesRequest{}
+// shouldLog checks the below conditions to see whether it
+// should emit logs.
+//   - Have we either:
+//     -- Gone past the server.telemetry.hot_ranges_stats.interval
+//     setting since th last time we logged or...
+//     -- Have we seen a hot range with a CPU time per second
+//     over the required threshold?
+//   - IF neither of the above conditions match, return false
+func shouldLog(
+	enabled bool,
+	ranges []*serverpb.HotRangesResponseV2_HotRange,
+	lastLogged time.Time,
+	logInterval time.Duration,
+	timeSrc timeutil.TimeSource,
+) bool {
+	if maxCPU(ranges) > LogCPUThreshold {
+		return true
+	}
+
+	if timeSrc.Since(lastLogged) > logInterval {
```

Comment:
seems like this is a check that we can do in `s.shouldLog`  that can happen before making the RPC call to get hot ranges.

Reply:
Not a bad idea I think in the context of not pulling the ranges to the last second, it might be cleaner to merge these into a single func, let me know how it looks.